onnxruntime/core/providers/cpu/object_detection/non_max_suppression.cc

```diff
@@ -24,192 +25,171 @@ ONNX_OPERATOR_KERNEL_EX(
       // Filter by score_threshold_
-      std::priority_queue<ScoreIndexPair, std::deque<ScoreIndexPair>, decltype(LessCompare)> sorted_scores_with_index(LessCompare);
-      for (int64_t box_index = 0; box_index < num_boxes_; ++box_index) {
-        if (!has_score_threshold || (has_score_threshold && scores_data[box_score_offset + box_index] > score_threshold)) {
-          sorted_scores_with_index.emplace(ScoreIndexPair({scores_data[box_score_offset + box_index], box_index}));
+      std::priority_queue<ScoreIndexPair, std::deque<ScoreIndexPair>> sorted_scores_with_index;
+      const auto* class_scores = scores_data + box_score_offset;
+      if (pc.score_threshold_ != nullptr) {
+        for (int64_t box_index = 0; box_index < pc.num_boxes_; ++box_index, ++class_scores) {
+          if (*class_scores > score_threshold) {
+            sorted_scores_with_index.emplace(*class_scores, box_index);
+          }
+        }
+      } else {
+        for (int64_t box_index = 0; box_index < pc.num_boxes_; ++box_index, ++class_scores) {
+          sorted_scores_with_index.emplace(*class_scores, box_index);
         }
       }
 
       ScoreIndexPair next_top_score;
       std::vector<int64_t> selected_indicies_inside_class;
-      // Get the next box with top score, filter by iou_threshold_
+      // Get the next box with top score, filter by iou_threshold
       while (!sorted_scores_with_index.empty()) {
         next_top_score = sorted_scores_with_index.top();
         sorted_scores_with_index.pop();
 
         bool selected = true;
         // Check with existing selected boxes for this class, suppress if exceed the IOU (Intersection Over Union) threshold
-        for (int64_t selected_indicies_inside_clas : selected_indicies_inside_class) {
-          if (SuppressByIOU(boxes_data + box_offset, selected_indicies_inside_clas, next_top_score.index,
-                            iou_threshold)) {
+        for (int64_t selected_index : selected_indicies_inside_class) {
+          if (SuppressByIOU(boxes_data + box_offset, selected_index, next_top_score.index_,
+                            center_point_box, iou_threshold)) {
             selected = false;
             break;
           }
         }
 
         if (selected) {
-          if (max_output_boxes_per_class > 0 && static_cast<int64_t>(selected_indicies_inside_class.size()) >= max_output_boxes_per_class) {
+          if (max_output_boxes_per_class > 0 &&
+              static_cast<int64_t>(selected_indicies_inside_class.size()) >= max_output_boxes_per_class) {
             break;
           }
-          selected_indicies_inside_class.push_back(next_top_score.index);
-          tmp_selected_indices.emplace_back(batch_index, class_index, next_top_score.index);
+          selected_indicies_inside_class.push_back(next_top_score.index_);
+          selected_indices.emplace_back(batch_index, class_index, next_top_score.index_);
         }
```

Comment:
Prefer push_back? 

[https://abseil.io/tips/112](https://abseil.io/tips/112)

Reply:
I do no think this applies here. I have seen this article before many times and all it does is confuses people making them forgo the benefits of emplace. The contrived example that applies to a specific case should not be generalized. And no, not every emplace() can be rewirtten by push_back()

---
In reply to: [298064286](https://github.com/microsoft/onnxruntime/pull/1303#discussion_r298064286) [](ancestors = 298064286)